File: web/lib/normalize.ts
// Display-layer text normalization. This is the last line of defense for
// professional casing: it runs on whatever the DB holds, so titles, companies,
// degrees, and institutions read cleanly regardless of how a source captured
// them (lowercase scrapes, ALL-CAPS headlines, "kbre", "llc", "texas a&m").
//
// It mirrors pipeline/normalize.py::smart_title so the DB write path and the
// render path agree. The render path is authoritative for what the user sees.

// Minor words stay lowercase unless they open the string.
const MINOR = new Set([
  "a", "an", "the", "and", "but", "or", "for", "nor",
  "on", "at", "to", "by", "in", "of", "up", "as", "vs",
]);

// Roman-numeral suffixes read as uppercase ("iii" -> "III"). Single "i" is
// excluded — it's ambiguous and capitalizes to "I" anyway.
const ROMAN = new Set(["ii", "iii", "iv", "v", "vi", "vii", "viii", "ix", "x"]);

// Tokens that must render fully uppercase even when a source lowercased them.
// Curated for this finance dataset: credentials, corporate suffixes that are
// genuinely acronyms (LLC/LP — but NOT Inc./Corp./Ltd./Co. which read as words),
// C-suite, and recurring orgs that appear lowercased in scraped text.
const ACRONYMS = new Set([
  // credentials / degrees
  "cfa", "caia", "cpa", "cfp", "frm", "mba", "bba", "emba", "llm", "phd",
  // corporate suffixes that are acronyms
  "llc", "lp", "llp", "plc", "pllc", "lllp",
  // C-suite
  "ceo", "cfo", "coo", "cto", "cio", "cmo", "cro", "evp", "svp",
  // recurring orgs / finance terms seen lowercased in this data
  "kkr", "kbre", "trs", "lse", "ccmp", "reit", "etf", "ipo",
]);

const PUNCT = ".,;:()[]\"'";

interface TokenParts {
  prefix: string;
  core: string;
  suffix: string;
}

// Peel leading/trailing punctuation so we inspect the word itself, e.g.
// "(2004)," -> {prefix:"(", core:"2004", suffix:"),"}.
function splitToken(token: string): TokenParts {
  let start = 0;
  let end = token.length;
  while (start < end && PUNCT.includes(token[start])) start += 1;
  while (end > start && PUNCT.includes(token[end - 1])) end -= 1;
  return {
    prefix: token.slice(0, start),
    core: token.slice(start, end),
    suffix: token.slice(end),
  };
}

// Capitalize each "&"-separated segment so "a&m" -> "A&M", "r&d" -> "R&D",
// while a plain word capitalizes normally ("finance" -> "Finance").
function capitalizeWord(word: string): string {
  return word
    .split("&")
    .map((seg) => (seg ? seg[0].toUpperCase() + seg.slice(1).toLowerCase() : seg))
    .join("&");
}

function isAllCapsAcronym(core: string): boolean {
  return (
    core.length > 1 &&
    /^[A-Za-z]+$/.test(core) &&
    core === core.toUpperCase()
  );
}

function isAlreadyMixedCase(core: string): boolean {
  // An uppercase letter after the first char means the source set it
  // deliberately (McCallum, DeVos, "A&M").
  return /[A-Z]/.test(core.slice(1));
}

/**
 * Title-case a string with investment-domain awareness. Idempotent.
 *
 * Priority: all-caps acronyms and already-mixed-case tokens are preserved;
 * curated acronyms and roman numerals go uppercase; minor words stay lowercase
 * unless first; everything else is capitalized (&-aware).
 *
 * Never pass verbatim source quotes here — they must stay untouched.
 */
export function smartTitle(value: string | null | undefined): string {
  if (!value) return value ?? "";
  const tokens = value.trim().split(/\s+/);
  return tokens
    .map((token, i) => {
      const { prefix, core, suffix } = splitToken(token);
      if (!core) return token;
      if (isAllCapsAcronym(core)) return token;
      if (isAlreadyMixedCase(core)) return token;

      const lower = core.toLowerCase();
      if (ACRONYMS.has(lower)) return prefix + lower.toUpperCase() + suffix;
      if (ROMAN.has(lower)) return prefix + lower.toUpperCase() + suffix;
      if (i > 0 && MINOR.has(lower)) return prefix + lower + suffix;
      return prefix + capitalizeWord(lower) + suffix;
    })
    .join(" ");
}

File: pipeline/normalize.py
"""Claim normalization and deduplication for Phase 2.

Two responsibilities:
1. smart_title() — title-case claim values so "senior investment manager"
   becomes "Senior Investment Manager" and "KKR" stays "KKR".
2. digest_claims() — deduplicate a merged claim set (e.g. Firecrawl + PDL)
   before writing to the DB, keeping the highest-confidence version of any
   duplicate and dropping exact case-insensitive matches.

Applied just before replace_claims() so the DB always holds clean data
regardless of which sources contributed.
"""
from __future__ import annotations

from enrichment_store import ClaimRow

# Words that stay lowercase unless they are the first word in the string.
_MINOR = frozenset({
    "a", "an", "the", "and", "but", "or", "for", "nor",
    "on", "at", "to", "by", "in", "of", "up", "as", "vs",
})

# Roman-numeral suffixes common in professional titles and names ("Partner III",
# "Managing Director II"). Sourced lowercase, these must read as uppercase, not
# "Iii". Single "i" is excluded — it's ambiguous and capitalizes to "I" anyway.
_ROMAN_SUFFIXES = frozenset({
    "ii", "iii", "iv", "v", "vi", "vii", "viii", "ix", "x",
})

# Claim types whose values should be title-cased.
_TITLE_CASE_TYPES = frozenset({
    "current_title",
    "current_employer",
    "career_history",
    "education",
    "location",
})


def smart_title(value: str) -> str:
    """Title-case a string with investment-domain awareness.

    Rules (in priority order):
    - All-caps tokens (KKR, TRS, LLC, LP, PE, CFO, CEO…) are preserved.
    - Mixed-case tokens already set by the source (McDonald, DeVos) are
      preserved.
    - Minor words (of, at, the, and…) are lowercased unless they are the
      first token.
    - Everything else is capitalized.

    The *quote* field is never passed here — it must stay verbatim.
    """
    if not value:
        return value
    tokens = value.strip().split()
    result: list[str] = []
    for i, tok in enumerate(tokens):
        # Strip leading/trailing punctuation to inspect the word itself.
        stripped = tok.strip(".,;:()[]\"'")
        prefix = tok[: len(tok) - len(tok.lstrip(".,;:()[]\"'"))]
        suffix = tok[len(tok.rstrip(".,;:()[]\"'")):]

        if not stripped:
            result.append(tok)
            continue

        # Preserve all-caps acronyms (>1 char, all uppercase letters/digits).
        if stripped.isupper() and len(stripped) > 1 and stripped.isalpha():
            result.append(tok)
            continue

        # Preserve already-mixed-case words (e.g. "McCallum", "DeVos").
        if any(c.isupper() for c in stripped[1:]):
            result.append(tok)
            continue

        word_lower = stripped.lower()

        # Roman-numeral suffixes read as uppercase ("iii" -> "III").
        if word_lower in _ROMAN_SUFFIXES:
            result.append(prefix + word_lower.upper() + suffix)
            continue

        # Minor words stay lowercase unless they open the string.
        if i > 0 and word_lower in _MINOR:
            result.append(prefix + word_lower + suffix)
        else:
            result.append(prefix + word_lower.capitalize() + suffix)

    return " ".join(result)


def _normalize_value(claim_type: str, value: str) -> str:
    """Apply smart_title to claim types that display as professional titles."""
    if claim_type in _TITLE_CASE_TYPES:
        return smart_title(value)
    return value


def normalize_claim(claim: ClaimRow) -> ClaimRow:
    """Return a new ClaimRow with the value title-cased where appropriate.
    The quote field is left untouched — it is a verbatim source excerpt."""
    normed = _normalize_value(claim.claim_type, claim.value)
    if normed == claim.value:
        return claim
    return ClaimRow(
        claim_type=claim.claim_type,
        value=normed,
        source_url=claim.source_url,
        quote=claim.quote,          # verbatim — never mutated
        confidence=claim.confidence,
        extraction_method=claim.extraction_method,
    )


def _confidence(claim: ClaimRow) -> float:
    """Confidence as a sortable float; a missing value floors to 0.0."""
    return claim.confidence if claim.confidence is not None else 0.0


def digest_claims(claims: list[ClaimRow]) -> list[ClaimRow]:
    """Normalize and deduplicate a merged claim set before persisting.

    Steps:
    1. Normalize case on all eligible claim values.
    2. Deduplicate: when two claims share the same (claim_type,
       case-insensitive value), keep the one with higher confidence.
       Ties go to the first seen (Firecrawl/Haiku typically comes first
       and has source-quoted evidence; PDL is secondary).

    news_mention claims are intentionally excluded from deduplication
    because multiple articles about the same person are all valid.
    """
    normalized = [normalize_claim(c) for c in claims]

    seen: dict[tuple[str, str], ClaimRow] = {}
    news: list[ClaimRow] = []

    for c in normalized:
        if c.claim_type == "news_mention":
            # Keep all news items — they are distinct articles, not duplicates.
            news.append(c)
            continue

        key = (c.claim_type, c.value.lower().strip())
        existing = seen.get(key)
        # Defensive: a malformed source could hand us a None confidence; treat it
        # as the lowest so it never wins a tie and never raises on comparison.
        if existing is None or _confidence(c) > _confidence(existing):
            seen[key] = c

    # Preserve insertion order: verified claims first, then news.
    return list(seen.values()) + news
